### PR TITLE
buffer: Optimize the layout of Slices in Buffer::OwnedImpl by removing subclassing and storing slice info directly in the SliceDeque

### PR DESCRIPTION
**Pull request by antoniovicente**
Commit Message:
buffer: Optimize the layout of Slices in Buffer::OwnedImpl by removing subclassing and storing slice info directly in the SliceDeque

Additional Description:
Risk Level: medium. Subtle changes to low level data structures. Manual implementation of move constructors.
Testing: unit
Docs Changes: n/a
Release Notes: n/a

### CODE REVIEW

File: test/common/buffer/buffer_test.cc [hunk 1]
```diff
@@ -14,21 +14,158 @@ namespace Envoy {
...
-  DummySlice(const std::string& data, const std::function<void()>& deletion_callback)
-      : Slice(0, data.size(), data.size()), deletion_callback_(deletion_callback) {
-    base_ = reinterpret_cast<uint8_t*>(const_cast<char*>(data.c_str()));
-  }
-  ~DummySlice() override {
-    if (deletion_callback_ != nullptr) {
-      deletion_callback_();
-    }
-  }
+TEST(SliceTest, OwnedMoveConstruction) {
```
Comment (antoniovicente) [inline test/common/buffer/buffer_test.cc:17]: TODO: change to TEST_P

File: source/common/buffer/buffer_impl.h [hunk 2]
```diff
@@ -30,22 +30,67 @@ namespace Buffer {
...
+        capacity_(size) {}
+
+  Slice(BufferFragment& fragment)
+      : storage_(nullptr), base_(static_cast<uint8_t*>(const_cast<void*>(fragment.data()))),
+        data_(0), reservable_(fragment.size()), capacity_(fragment.size()) {
+    addDrainTracker([&fragment]() { fragment.done(); });
+  }
+
+  Slice(Slice&& rhs) noexcept {
+    storage_ = std::move(rhs.storage_);
```
Comment (antoniovicente) [inline source/common/buffer/buffer_impl.h:50]: TODO: Use helper function to code shared between move constructor and move assignment?

**Review by ggreenway — CHANGES_REQUESTED** [review]
I like the approach. This will be much easier/smaller to review without the snake-to-camel case name change though.

/wait

File: source/common/buffer/buffer_impl.h [hunk 3]
```diff
@@ -30,22 +30,80 @@ namespace Buffer {
...
-    RELEASE_ASSERT(isMutable(), "Not allowed to call getMutableData if slice is immutable");
-    return {base_ + data_, static_cast<absl::Span<uint8_t>::size_type>(reservable_ - data_)};
+  /**
+   * Create an empty mutable Slice that owns its storage.
+   * @param min_capacity number of bytes of space the slice should have. Actual capacity is rounded
+   * up to the next multiple of 4kb.
+   */
+  Slice(uint64_t min_capacity)
+      : capacity_(sliceSize(min_capacity)), storage_(new uint8_t[capacity_]),
+        base_(storage_.get()) {}
```
[the remark this answers is not in the document]
Comment (antoniovicente) [reply source/common/buffer/buffer_impl.h]: Done.  An earlier clang-tidy warning motivated me to move initialization to the member definition.

File: source/common/buffer/buffer_impl.h [hunk 4]
```diff
@@ -233,74 +286,53 @@ class Slice : public SliceData {
...
+   * @param data_size the minimum amount of data the slice must be able to store, in bytes.
+   * @return a recommended slice size, in bytes.
+   */
+  static uint64_t sliceSize(uint64_t data_size) {
+    static constexpr uint64_t PageSize = 4096;
+    const uint64_t num_pages = (data_size + PageSize - 1) / PageSize;
+    return num_pages * PageSize;
+  }
+
+  /** Total number of bytes in the slice */
```
[the remark this answers is not in the document]
Comment (antoniovicente) [reply source/common/buffer/buffer_impl.h]: Updated comments.

File: source/common/buffer/buffer_impl.h [hunk 5]
```diff
@@ -332,14 +364,14 @@ class SliceDeque {
     return *this;
   }
 
-  void emplace_back(SlicePtr&& slice) {
+  void emplaceBack(Slice&& slice) {
```
[the remark this answers is not in the document]
Comment (antoniovicente) [reply source/common/buffer/buffer_impl.h]: Rolled back the change and added NOLINTs.

File: source/common/buffer/buffer_impl.h [hunk 6]
```diff
@@ -349,31 +381,37 @@ class SliceDeque {
...
-  const SlicePtr& back() const { return ring_[internalIndex(size_ - 1)]; }
+  Slice& front() { return ring_[start_]; }
+  const Slice& front() const { return ring_[start_]; }
+  Slice& back() { return ring_[internalIndex(size_ - 1)]; }
+  const Slice& back() const { return ring_[internalIndex(size_ - 1)]; }
 
-  SlicePtr& operator[](size_t i) { return ring_[internalIndex(i)]; }
-  const SlicePtr& operator[](size_t i) const { return ring_[internalIndex(i)]; }
+  Slice& operator[](size_t i) {
+    ASSERT(size_ > 0);
```
[the remark this answers is not in the document]
Comment (antoniovicente) [reply source/common/buffer/buffer_impl.h]: Changed to ASSERT(!empty());

A more general check would be:
ASSERT(i < size_);

**Review by ggreenway — COMMENTED** [review]
Mostly LGTM

File: source/common/buffer/buffer_impl.cc [hunk 7]
```diff
@@ -196,16 +196,17 @@ RawSliceVector OwnedImpl::getRawSlices(absl::optional<uint64_t> max_slices) cons
     // there is currently no max size validation.
     // TODO(antoniovicente) Set realistic limits on the max size of BufferSlice and consider use of
     // size_t instead of uint64_t in the Slice interface.
-    raw_slices.emplace_back(RawSlice{slice->data(), static_cast<size_t>(slice->dataSize())});
+    raw_slices.emplace_back(
+        RawSlice{const_cast<uint8_t*>(slice.data()), static_cast<size_t>(slice.dataSize())});
```
[the remark this answers is not in the document]
Comment (antoniovicente) [reply source/common/buffer/buffer_impl.cc:200]: The introduction of ConstRawSlice in parallel to RawSlice will require API across multiple interfaces.  Some examples include arguments to IoHandle::writev which should use ConstRawSlice, while readv should continue using RawSlice.  Some const casts would be needed when you have ConstRawSlice and are feeding that input into C APIs like nghttp2_hd_inflate_hd2 in MetadataDecoder::decodeMetadataPayloadUsingNghttp2

This PR already feels huge.  Happy to try the change as a followup if you'ld like.

[the remark this answers is not in the document]
Comment (ggreenway) [reply source/common/buffer/buffer_impl.cc:200]: Yeah, agree we shouldn't expand the scope that much in this PR.

**Review by ggreenway — APPROVED** [review]

**Pull request merged**